library/tiqr/tests/OcraTest.php

```diff
@@ -99,4 +255,40 @@ public function testTruncate() {
         $this->assertEquals( 0, (int)OCRA::_oath_truncate(str_repeat("00", 20), 6) );
     }
 
+    /**
+     * @dataProvider HexToBinTestProvider
+     */
+    function testHexToBin($expected, string $hex, int $maxBytes, string $parameterName) {
+        // Use reflection to access the private static function OCRA::_hexStr2Bytes(string $hex, int $maxBytes, string $parameterName) : string
+
+        $classOCRA = new ReflectionClass('OCRA');
+        $_hexStr2Bytes = $classOCRA->getMethod("_hexStr2Bytes");
+        $_hexStr2Bytes->setAccessible(true);
+
+        if ( $expected instanceof Exception ) {
+            $this->expectExceptionObject($expected);
+            $_hexStr2Bytes->invokeArgs(NULL, array($hex, $maxBytes, $parameterName));
+        }
+        else {
+            $this->assertSame($expected, $_hexStr2Bytes->invokeArgs(NULL, array($hex, $maxBytes, $parameterName)));
+        }
+    }
+
+    // For testHexToBin
+    public function HexToBinTestProvider(): array
+    {
+        return [
+            // [ expected return value or exception, hex, maxBytes, parameterName]
+            ['', '', 0, 'dummy'], // empty string => empty result
```

Comment:
When building a data provider, a nice addition is to make the test input an associative array. The key of the entries are used in the test output. That way you can more easily identify which of the test variants passed or failed.

Example:

```php
return [
    'Empty string => empty result' => ['', '', 0, 'dummy'],
   'Case does not matter' =>["\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\xaa\xbb\xcc\xdd\xee\xff", '00010203040506070809aAbBcCdDeEfF', 16, ''],
]
```

Reply:
This is cool @MKodde ! I didn't know this..

Reply:
Good tip @MKodde, thanks!